dummyfs/dummyfs.c

```diff
@@ -511,470 +639,509 @@ int dummyfs_create(void *ctx, oid_t *dir, const char *name, oid_t *oid, unsigned
+	if (o == NULL) {
 		return -ENOENT;
+	}
 
-	if ((ret = object_remove(fs, o)) == EOK) {
+	int ret = dummyfs_object_remove(fs, o);
+	if (ret == 0) {
 		if (S_ISREG(o->mode)) {
-			object_lock(fs, o);
-			dummyfs_truncate_internal(fs, o, 0);
-			object_unlock(fs, o);
+			(void)_dummyfs_truncateObject(fs, o, 0);
 		}
 		else if (S_ISDIR(o->mode)) {
-			/* check if mountpoint */
-			if (o->oid.id != o->dev.id) {
-				dev_destroy(fs, &o->dev);
-			}
-			dir_destroy(fs, o);
-		}
-		else if (S_ISCHR(o->mode) || S_ISBLK(o->mode) || S_ISFIFO(o->mode)) {
-			dev_destroy(fs, &o->dev);
+			dummyfs_dir_destroy(fs, o);
 		}
 		else if (o->mode == OBJECT_MODE_MEM) {
-#ifndef NOMMU
-			munmap((void *)((uintptr_t)o->chunks->data & ~(_PAGE_SIZE - 1)), (o->size + (_PAGE_SIZE - 1)) & ~(_PAGE_SIZE - 1));
-#endif
-			free(o->chunks);
+			munmap(o->data, (o->size + (_PAGE_SIZE - 1)) & ~(_PAGE_SIZE - 1));
+			o->data = NULL;
 		}
-		dummyfs_decsz(fs, sizeof(dummyfs_object_t));
-		free(o);
+		dummyfs_free(fs, o, sizeof(dummyfs_object_t));
 	}
 
 	return ret;
 }
 
 
+int dummyfs_destroy(void *ctx, oid_t *oid)
+{
+	TRACE;
+	dummyfs_t *fs = (dummyfs_t *)ctx;
+	mutexLock(fs->mutex);
+	int ret = _dummyfs_destroy(fs, oid);
+	mutexUnlock(fs->mutex);
+	return ret;
+}
+
+
 int dummyfs_readdir(void *ctx, oid_t *dir, offs_t offs, struct dirent *dent, unsigned int size)
 {
+	TRACE;
 	dummyfs_t *fs = (dummyfs_t *)ctx;
-	dummyfs_object_t *d;
-	dummyfs_dirent_t *ei;
-	offs_t diroffs = 0;
-	int ret = -ENOENT;
 
-	if (dummyfs_device(fs, dir))
+	if (dummyfs_isDevice(fs, dir) != 0) {
 		return -EINVAL;
+	}
 
-	d = object_get(fs, dir->id);
-
-	if (d == NULL)
+	mutexLock(fs->mutex);
+	dummyfs_object_t *d = dummyfs_object_get(fs, dir);
+	if (d == NULL) {
+		mutexUnlock(fs->mutex);
 		return -ENOENT;
+	}
 
-	if (!S_ISDIR(d->mode)) {
-		object_put(fs, d);
+	if (!S_ISDIR(d->mode) || (offs < 0)) {
+		dummyfs_object_put(fs, d);
+		mutexUnlock(fs->mutex);
 		return -EINVAL;
 	}
 
-	object_lock(fs, d);
-
-	if ((ei = d->entries) == NULL) {
-		object_unlock(fs, d);
-		object_put(fs, d);
-		return -EINVAL;
+	dummyfs_dirent_t *ei;
+	dummyfs_dirent_t *etree;
+	off_t diroffs = 0;
+	if ((d->dir.hint.entry != NULL) && (offs >= d->dir.hint.offs)) {
+		diroffs = d->dir.hint.offs;
+		ei = d->dir.hint.entry;
+		etree = ei;
+		while (etree->prev != NULL) {
+			etree = etree->prev;
+		}
+	}
+	else {
+		ei = lib_treeof(dummyfs_dirent_t, linkage, lib_rbMinimum(d->dir.tree.root));
+		etree = ei;
 	}
-	dent->d_reclen = 0;
-	do {
-		if (diroffs >= offs) {
-			if ((sizeof(struct dirent) + ei->len + 1) > size) {
-				object_unlock(fs, d);
-				object_put(fs, d);
-				return 	-EINVAL;
-			}
-			if (ei->deleted) {
-				ei = ei->next;
-				dent->d_reclen++;
-				continue;
-			}
 
-			dent->d_ino = ei->oid.id;
-			dent->d_reclen++;
-			dent->d_namlen = ei->len;
-			dent->d_type = ei->type;
-			strcpy(dent->d_name, ei->name);
+	while ((diroffs < offs) && (ei != NULL)) {
+		diroffs += ei->len;
```

Comment:
If entry is unlinked in between readdir calls we can end up skipping over few valid entries

Reply:
Yup, and there's nothing that can be done about that (apart from making snapshot)

Reply:
It's not POSIX compliant then
```
The type DIR, which is defined in the [<dirent.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/dirent.h.html) header, represents a directory stream, which is an ordered sequence of ALL the directory entries in a particular directory. Directory entries represent files; files may be removed from a directory or added to a directory asynchronously to the operation of readdir().
```
only exception is 
```
If a file is removed from or added to the directory after the most recent call to [opendir()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/opendir.html) or [rewinddir()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/rewinddir.html), whether a subsequent call to readdir() returns an entry for that file is unspecified.
```

Reply:
Let's ignore this problem for now, I've created an issue as this require some architecture changes

https://github.com/phoenix-rtos/phoenix-rtos-project/issues/900